controllers/securitygroup/securitygroup_controller.go

```diff
@@ -69,71 +71,203 @@ const (
 type SecurityGroupReconciler struct {
 	client.Client
 	Scheme                          *runtime.Scheme
-	log                             logr.Logger
 	Recorder                        record.EventRecorder
 	NewEC2ClientFactory             func(cfg aws.Config) ec2.EC2Client
 	NewAutoScalingClientFactory     func(cfg aws.Config) autoscaling.AutoScalingClient
 	NewOceanCloudProviderAWSFactory func() spot.OceanClient
 }
 
+type SecurityGroupReconciliation struct {
+	SecurityGroupReconciler
+	log                logr.Logger
+	start              time.Time
+	ec2Client          ec2.EC2Client
+	asgClient          autoscaling.AutoScalingClient
+	sg                 *securitygroupv1alpha1.SecurityGroup
+	kmp                *kinfrastructurev1alpha1.KopsMachinePool
+	kcp                *kcontrolplanev1alpha1.KopsControlPlane
+	providerConfigName string
+}
+
 //+kubebuilder:rbac:groups=ec2.aws.wildlife.io,resources=securitygroups,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=ec2.aws.wildlife.io,resources=securitygroups/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=ec2.aws.wildlife.io,resources=securitygroups/finalizers,verbs=update
 //+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=kopsmachinepools,verbs=get;list;watch
 //+kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=kopscontrolplanes,verbs=get;list;watch
 //+kubebuilder:rbac:groups=ec2.aws.crossplane.io,resources=securitygroups,verbs=get;list;watch;create;update;patch;delete
 
-func (r *SecurityGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
-	start := time.Now()
-	r.log = ctrl.LoggerFrom(ctx)
+func DefaultReconciler(mgr manager.Manager) *SecurityGroupReconciler {
+	return &SecurityGroupReconciler{
+		Client:                          mgr.GetClient(),
+		Scheme:                          mgr.GetScheme(),
+		Recorder:                        mgr.GetEventRecorderFor("securityGroup-controller"),
+		NewEC2ClientFactory:             ec2.NewEC2Client,
+		NewAutoScalingClientFactory:     autoscaling.NewAutoScalingClient,
+		NewOceanCloudProviderAWSFactory: spot.NewOceanCloudProviderAWS,
+	}
+}
+
+func awsConfigForCredential(ctx context.Context, region string, accessKey string, secretAccessKey string) (aws.Config, error) {
```

Comment:
Since we are using this function for the Clustermesh controller, we can move this to separated package and reuse it